src/i18n.ts

```diff
@@ -0,0 +1,280 @@
+  );
+}
+
+export class I18n<C extends Context = Context> {
+  private config: I18nConfig<C>;
+  readonly fluent: Fluent;
+  readonly locales = new Array<string>();
+
+  constructor(config: Partial<I18nConfig>) {
+    this.config = {
+      defaultLocale: "en",
+      fluentBundleOptions: { functions: {} },
+      ...config,
+    };
+
+    if (
+      this.config.fluentBundleOptions &&
+      this.config.fluentBundleOptions.functions
+    ) {
+      this.config.fluentBundleOptions.functions["CTX"] = CTX;
+    } else {
+      this.config.fluentBundleOptions = {
+        ...this.config.fluentBundleOptions,
+        functions: { CTX },
+      };
+    }
+
+    this.fluent = new Fluent(this.config.fluentOptions);
+
+    if (config.directory) {
+      this.loadLocalesDir(config.directory);
+    }
+  }
+
+  /**
+   * Loads locales from the specified directory and registers them in the Fluent instance.
+   * @param directory Path to the directory to look for the translation files.
+   */
+  loadLocalesDir(directory: string): void {
+    for (const file of readLocalesDir(directory)) {
+      const path = resolve(directory, file);
+      const locale = file.substring(0, file.lastIndexOf("."));
+
+      this.loadLocale(locale, {
+        filePath: path,
+        bundleOptions: this.config.fluentBundleOptions,
+      });
+    }
+  }
+
+  /**
+   * Registers a locale in the Fluent instance based on the provided options.
+   * @param locale Locale ID
+   * @param options Options to specify the source and behavior of the translation
+   */
+  loadLocale(
+    locale: LocaleId,
+    options: {
+      filePath?: string;
+      source?: string;
+      isDefault?: boolean;
+      bundleOptions?: FluentBundleOptions;
+    },
+  ): void {
+    this.fluent.addTranslation({
+      locales: locale,
+      isDefault: locale === this.config.defaultLocale,
+      bundleOptions: this.config.fluentBundleOptions,
+      ...options,
+    });
+
+    this.locales.push(locale);
+  }
+
+  /** Gets an array of translations for a given key. */
+  t(key: string): Array<string>;
+  /** Gets a message by its key from the specified locale. */
+  t(locale: LocaleId, key: string, context?: TranslationContext): string;
+  t(
+    keyOrLocale: string | LocaleId,
+    key?: string,
+    context?: TranslationContext,
+  ): string | Array<string> {
+    if (keyOrLocale && key) {
+      return this.fluent.translate(keyOrLocale, key, context);
+    }
+
+    const translations = new Array<string>();
+    for (const locale of this.locales) {
+      const message = this.translate(locale, keyOrLocale, context);
+      translations.push(message);
+    }
+
+    return translations;
+  }
+
+  /** Gets a message by its key from the specified locale. */
+  translate(key: string): Array<string>;
+  translate(
+    locale: LocaleId,
+    key: string,
+    context?: TranslationContext,
+  ): string;
+  translate(
+    keyOrLocale: string,
```

Comment:
I don't think you got the point here. When only key is passed, this method should return translated messages as an array of strings. So, it can be used in `bot.hears(i18n.t("key"))` like situations. But I think you got a point in moving the `locale` parameter and not having this method overloading.

I am re-thinking about this method right now, and moving the above mentioned hears-matching stuff in to something like `match`. And probably taking your suggestion on returning locales as an object.

Reply:
@roj1512 Any suggestion about this?

Reply:
@ngdangtu-vn You suggested:
```ts
i18n.t("key", ["en", "de", "ru"])
```
which should be returning
```ts
{ en: "EN", de: "DE", ru: "RU" }
```

What do you think of preserving the order instead of mapping the translations? Like returning this:

```ts
["EN", "DE", "RU"]
```

Reply:
I didn't think throughly about the use case of this. Aside `bot.hears` case, do we have other cases for flat array? I will do more research about this in the late day.

Reply:
> I didn't think throughly about the use case of this.

I don't think we need a feat if it does not have a proper use case related to grammY. And I don't think aside from `bot.hears` there are really useful use cases. I might be wrong, but I think there isn't something useful. But I don't think we have to worry about that yet.

Reply:
You're right, let's not try to cover all cases. So we will go with the flat array but no overloading method, right?

Reply:
Why no method overloading? 🤔

Reply:
For this case, I think it is over complicated.

Reply:
No it is not. I've tried the normal way and there were type issues in certain cases. That's why I came back to overloading.

Reply:
Ok, I got it. I'll start to review the latest code at the end of this week. :D